### ISSUE TITLE
Can sregistry be have keystore end point be added? 

### ISSUE THREAD

**Issue by asmi10** [singularityhub/sregistry#203]
Hi!

I have setup sregistry and would like to store encrypted images in sregistry, can we have keystore endpoint be added in sregistry for for managing keys?

**Comment by vsoch** [comment]
This was asked previously https://github.com/singularityhub/sregistry/issues/169 and the answer is yes! If you can walk me through what kinds of interactions you are looking for, I can start to think about it. 

A quick note that #172 is going to add GitHub + Google Cloud Build, and it's quite a big PR, so I'd want  to get it merged before working on this. I'm having someone review it more thoroughly on Monday so we could get started on it after that.

**Comment by vsoch** [comment]
hey @asmi10 not sure if you had an update or could share your vision for this, but the testing period for Google Cloud Build + GitHub is open, until the end of next week. You can participate here https://twitter.com/vsoch/status/1144361770661928962

**Comment by asmi10** [comment]
Hello vsoch,

Sorry, I was busy with some other important work. 
Currently I am using sregistry setup for storing sif images, and looking for functionality to store signed images in sregistry with public key. I went through #169 and it seemed you made some good progress there.

I see in #169 you were looking for some info about matching public key extracted from container  header with public key stored in user's cache. In singularity versions >= 3.0 we have sif tool incorporated in singularity, so the fingerprints for key in user's cache can be matched with signature object descriptor fingerprint. 

musxcnt004:/localdisk/images # sg key list
Public key listing (/root/.singularity/sypgp/pgp-public):

0) U: asmita thapliyal () <asmita.thapliyal@intel.com>
   C: 2019-09-30 10:07:57 +0200 CEST
   F: B8FF358C2D13B259CD9237A2DEDD08435C737BC1
   L: 4096
   --------

# sg sif list sles.sif
Container id: f5549d3d-25ad-4253-b22c-a7129aa149c4
Created on:   2019-09-30 10:07:17 +0200 CEST
Modified on:  2019-09-30 10:09:46 +0200 CEST
----------------------------------------------------
Descriptor list:
ID   |GROUP   |LINK    |SIF POSITION (start-end)  |TYPE
------------------------------------------------------------------------------
1    |1       |NONE    |32768-32802               |Def.FILE
2    |1       |NONE    |36864-29855744            |FS (Squashfs/*System/amd64)
3    |1       |2       |29855744-29856699         |Signature (SHA384)
4    |1       |1       |29859840-29860795         |Signature (SHA384)

musxcnt004:/localdisk/images # sg sif info 3 sles.sif
Descr slot#: 2
  Datatype:  Signature
  ID:        3
  Used:      true
  Groupid:   1
  Link:      2
  Fileoff:   29855744
  Filelen:   955
  Ctime:     2019-09-30 10:09:46 +0200 CEST
  Mtime:     2019-09-30 10:09:46 +0200 CEST
  UID:       0
  Gid:       0
  Name:      part-signature
  Hashtype:  SHA384
  Entity:    B8FF358C2D13B259CD9237A2DEDD08435C737BC1

**Comment by vsoch** [comment]
hey @asmi10 - thanks for the detail on sif tools! I'm definitely aware of how we can use siftool, SIF Python, and now even [a browser](https://vsoch.github.io/2019/singularity-web/) to derive this information - what I'm interested to understand is what the interaction would look like between a user and the registry. For example, it might be something like:

 - user enables keystore plugin for sregistry server
 - upon image upload, all keys are automatically added to the keystore (and what does this mean? The unique ID and the signature? The signatures should be linked to the images?
 - when a user downloads an image - do they also interact with the keystore?

You can assume I've never used a keystore and I don't know how it should work. For example, why should we have keys in the server if the user can just download a container and then see the keys that it's signed with? What are all the interactions a user can have with a keystore?

Thanks in advance! This should be fairly do-able, but first I need to understand all the details :)

**Comment by asmi10** [comment]
Thanks a lot for helping me here!
Basically it should work like sylabs keystore endpoint: https://cloud.sylabs.io/keystore.
 
>  While image upload only images should be uploaded to registry, users can generate keys via cmd: singularity key newpair, and could push it to keystore via singularity key push <ID>. While push it should check the user identity, in below example I have used different email id so push got failed.
 
/localdisk/images # sg key newpair
Enter your name (e.g., John Doe) : asmita thapliyal
Enter your email address (e.g., john.doe@example.com) : asmita.thapliyal@yahoo.com
Enter optional comment (e.g., development keys) :
Enter a passphrase :
Retype your passphrase :
Would you like to push it to the keystore? [Y,n] y
Generating Entity and OpenPGP Key Pair... done
Failed to push newly created key to keystore: key server did not accept PGP key: entity B41A8DAFFC392C69 does not contain identity with email 'asmita.thapliyal@gmail.com' (400 Bad Request)
 
 
> While image download it should check the image signature, if no matching keys found in user's cache then list the warning like below . 
:/infrastructure # singularity pull library://asmita/default/panda:12.sif
INFO:    Downloading library image
 28.48 MiB / 28.48 MiB [=============================================================================================================================================] 100.00% 6.29 MiB/s 4s
WARNING: Container might not be trusted; run 'singularity verify panda_12.sif.sif' to show who signed it
INFO:    Download complete: panda_12.sif.sif
 
If user runs verify command then keystore interaction is needed for verifying the signature.
musxtestvm006:/infrastructure # singularity verify panda_12.sif.sif
Container is signed by 1 key(s):
 
Verifying signature F: 3BAA1E4A42D9B99727462A03689F239612B5D219:
[REMOTE]  asmita thapliyal <asmita.thapliyal@gmail.com>
[OK]      Data integrity verified
 
INFO:    Container verified: panda_12.sif.sif
 
If users cache is already having matching keys of container then it should be verified automatically.
 
:/infrastructure # singularity key search asmita
Showing 2 results
 
KEY ID    BITS  NAME/EMAIL
7FD11695  4096  asmita thapliyal <asmita.thapliyal@gmail.com>
 
12B5D219  4096  asmita thapliyal <asmita.thapliyal@gmail.com>
 
:/infrastructure # singularity key pull 12B5D219
1 key(s) added to keyring of trust /root/.singularity/sypgp/pgp-public
musxtestvm006:/infrastructure # rm panda_12.sif.sif
musxtestvm006:/infrastructure # singularity pull library://asmita/default/panda:12.sif
INFO:    Container is trusted - run 'singularity key list' to list your trusted keys
INFO:    Download complete: panda_12.sif.sif
 
Keystore interaction is needed in below cases
 
•	singularity key search
•	singularity key push <ID>
•	singularity key pull <ID>
•	singularity verify <container>
 
Basically signing image provides a trusted method for Singularity users to share containers. It ensures a bit-for-bit reproduction of the original container as the author intended it. You may compare it with rpm package signatures to implement integrity checks for packages. I am trying to implement some kind of control so that images are signed with only few trusted parties, or may be implement some kind of revoke method if it is not signed with trusted parties. This can be done later, first priority is to add keystore endpoint and implement user interaction with it.

**Comment by vsoch** [comment]
@asmi10 I agree that these are the endpoints that would need to be implemented - all of these commands allow specifying a url for "a key server" : 

 - singularity key search
 - singularity key push
 - singularity key pull
 - singularity key verify

I think what we'd want to do (akin to how sregistry serves a library endpoint) is to serve the same key endpoint as Sylabs, that way to not need to also provide a custom tool to interact with it. For example, we can see from `singularity key pull` that you can customize the url:

```
Download a public key from a key server

Usage:
  singularity key pull [pull options...] <fingerprint>

Description:
  The 'key pull' command allows you to connect to a key server look for and 
  download a public key. Key rings are stored into (e.g., 
  $HOME/.singularity/sypgp).

Options:
  -h, --help         help for pull
  -u, --url string   specify the key server URL (default
                     "https://keys.sylabs.io")
```
And then the user would need to specify --url to say use a different key registry. Do you think that folks would do that?

If so, then we need to think about implementation. It's either the case that they are using some key server out of the box, or that it's a custom endpoint (akin to their library). Looking through the code, I'm seeing a [key server client](https://github.com/sylabs/singularity/blob/master/pkg/sypgp/sypgp.go#L815) instantiated that does a push (or a general action) and it looks like most of the functions are [here](https://github.com/sylabs/scs-key-client/blob/master/client/pks.go). My question is basically - is this some standard implementation, and one we could easily find a [python server for](https://github.com/mugwort-rc/django-pgpdb)?  Or something entirely custom?

**Comment by vsoch** [comment]
Making some progress! We can see the urls being queried by getting it wrong:

```
$ singularity key search --url http://boomshakala tacos
ERROR:   search failed: failed to get key: Get http://boomshakala/pks/lookup?fingerprint=on&op=index&options=mr&search=tacos&x-pagesize=256: dial tcp: lookup boomshakala on 127.0.0.53:53: server misbehaving
```
And add (push)

```
$ singularity key push --url http://boomshakala CFA6763B11637E52404A25F5DE565315F5198C71ERROR:   push failed: key server did not accept PGP key: Post http://boomshakala/pks/add: dial tcp: lookup boomshakala on 127.0.0.53:53: server misbehaving
```

**Comment by vsoch** [comment]
omg I think I got it working. lol, why was that so easy? Give me some time to clean things up, write up docs, and I'll hand you something to test. I'll implement it as a plugin for sregistry so a user isn't forced to add a keystore.

**Comment by asmi10** [comment]
Hi vsoch,

Thanks a lot for providing keystore function!
I was just trying it out after enabling pgp plugin, but I couldn't see any pks tab in sregistry web page. Also it seems push function only works with TLS, so I tried https but found below error. I am using using singularity version 3.4.1 master branch.

uwsgi_1      | Not Found: /pks/add/                                    │/sregistrytest.hpc.caas.intel.com 248A205C1795F62CF6177C774438D434EDF9F541
uwsgi_1      | [pid: 36|app: 0|req: 1/19] 10.216.137.221 () {38 vars in│ERROR:   push failed: key server did not accept PGP key: 404 Not Found
 1155 bytes} [Mon Oct  7 14:05:40 2019] GET /pks/add/ => generated 652 │musxtestvm006:~ # /infrastructure/sing/bin/singularity key push --url https:/
bytes in 916 msecs (HTTP/1.1 404) 3 headers in 116 bytes (1 switches on│/sregistrytest.hpc.caas.intel.com 248A205C1795F62CF6177C774438D434EDF9F541
 core 1)                                                               │ERROR:   push failed: key server did not accept PGP key: 404 Not Found
